crates/misc/component-fuzz-util/src/lib.rs

```diff
@@ -0,0 +1,761 @@
+//! parameters to the imported one and forwards the result back to the caller.  This serves to excercise Wasmtime's
+//! lifting and lowering code and verify the values remain intact during both processes.
+
+use arbitrary::{Arbitrary, Unstructured};
+use proc_macro2::{Ident, TokenStream};
+use quote::{format_ident, quote};
+use std::fmt::{self, Debug, Write};
+use std::iter;
+use std::ops::Deref;
+use wasmtime_component_util::{DiscriminantSize, FlagsSize, REALLOC_AND_FREE};
+
+const MAX_FLAT_PARAMS: usize = 16;
+const MAX_FLAT_RESULTS: usize = 1;
+const MAX_ARITY: usize = 5;
+
+/// The name of the imported host function which the generated component will call
+pub const IMPORT_FUNCTION: &str = "echo";
+
+/// The name of the exported guest function which the host should call
+pub const EXPORT_FUNCTION: &str = "echo";
+
+/// Maximum length of an arbitrary tuple type.  As of this writing, the `wasmtime::component::func::typed` module
+/// only implements the `ComponentType` trait for tuples up to this length.
+const MAX_TUPLE_LENGTH: usize = 16;
+
+#[derive(Copy, Clone, PartialEq, Eq)]
+enum CoreType {
+    I32,
+    I64,
+    F32,
+    F64,
+}
+
+impl CoreType {
+    /// This is the `join` operation specified in [the canonical
+    /// ABI](https://github.com/WebAssembly/component-model/blob/main/design/mvp/CanonicalABI.md#flattening) for
+    /// variant types.
+    fn join(self, other: Self) -> Self {
+        match (self, other) {
+            _ if self == other => self,
+            (Self::I32, Self::F32) | (Self::F32, Self::I32) => Self::I32,
+            _ => Self::I64,
+        }
+    }
+}
+
+impl fmt::Display for CoreType {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            Self::I32 => f.write_str("i32"),
+            Self::I64 => f.write_str("i64"),
+            Self::F32 => f.write_str("f32"),
+            Self::F64 => f.write_str("f64"),
+        }
+    }
+}
+
+/// Wraps a `Box<[T]>` and provides an `Arbitrary` implementation that always generates non-empty slices
+#[derive(Debug)]
+pub struct NonEmptyArray<T>(Box<[T]>);
+
+impl<'a, T: Arbitrary<'a>> Arbitrary<'a> for NonEmptyArray<T> {
+    fn arbitrary(input: &mut Unstructured<'a>) -> arbitrary::Result<Self> {
+        Ok(Self(
+            iter::once(input.arbitrary())
+                .chain(input.arbitrary_iter()?)
+                .collect::<arbitrary::Result<_>>()?,
+        ))
+    }
+}
+
+/// Wraps a `Box<[T]>` and provides an `Arbitrary` implementation that always generates non-empty slices of length
+/// less than or equal to the longest tuple for which Wasmtime generates a `ComponentType` impl
+#[derive(Debug)]
+pub struct TupleArray<T>(Box<[T]>);
+
+impl<'a, T: Arbitrary<'a>> Arbitrary<'a> for TupleArray<T> {
+    fn arbitrary(input: &mut Unstructured<'a>) -> arbitrary::Result<Self> {
+        Ok(Self(
+            iter::once(input.arbitrary())
```

Comment:
Similar to the cases below I think that we'll want to remove this to allow testing zero-sized tuples as well.

Reply:
I was thinking a zero-sized tuple was just a `unit` and thus redundant.  Should we also allow zero-sized records, variants, unions, etc.?  I.e. should I remove the `NonEmptyArray` type entirely?

Reply:
We should allow  zero-size records, yeah, but variants/unions/enums all require at least one case so the `NonEmptyArray` there is correct. You're right that in Rust an empty tuple will translate to a `()`, but for fuzzing here I think it's best to exercise as many cases as possible.